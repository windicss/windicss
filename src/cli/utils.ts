import fs from 'fs';
import path from 'path';
import glob from 'glob';
import minimatch from 'minimatch';

export function globArray(patterns: string[], options?: glob.IOptions): string[] {
  const list: string[] = [];

  patterns.forEach(pattern => {
    if (pattern[0] === '!') {
      let i = list.length-1;
      while(i > -1) {
        if (!minimatch(list[i], pattern)) {
          list.splice(i, 1);
        }
        i--;
      }

    }
    else {
      const newList = glob.sync(pattern, options);
      newList.forEach(item => {
        if (list.indexOf(item)===-1) {
          list.push(item);
        }
      });
    }
  });

  return list;
}

export function getVersion(): string {
  // eslint-disable-next-line quotes
  return `__NAME__ __VERSION__`; // replace by rollup
}

export function fuzzy(content: string): string[] {
  return content.match(/([^<>"'`\s]*[^<>"'`\s:])|([^<>"'`\s.(){}[\]#=%]*[^<>"'`\s.(){}[\]#=%:])/g) ?? [];
}

export function generateTemplate(
  folder: string,
  outputPath = 'windi.css'
): { html: string; css: string } {
  if (!(fs.existsSync(folder) && fs.lstatSync(folder).isDirectory())) {
    fs.mkdirSync(folder);
    if (!fs.existsSync(folder))
      throw new Error(`Folder ${folder} creation failed.`);
  }
  folder = path.resolve(folder);
  const template = `<!DOCTYPE html>
  <html lang="en">

  <head>
      <meta charset="UTF-8">
      <meta name="viewport" content="width=device-width, initial-scale=1.0">
      <title>${path.basename(folder)}</title>
      <link rel="stylesheet" href="${outputPath}">
  </head>

  <body class="bg-gray-100">
      <div class="container mx-auto flex flex-col justify-center items-center h-screen">
          <div class="lg:flex shadow rounded-lg">
              <div class="bg-blue-500 rounded-t-lg lg:rounded-tr-none lg:rounded-l-lg lg:w-4/12 py-4 h-full flex flex-col justify-center">
                  <div class="text-center tracking-wide">
                      <div class="text-white font-bold text-8xl lg:text-4xl">24</div>
                      <div class="text-white font-normal text-4xl pt-4 lg:pt-0 lg:text-2xl">Sept</div>
                  </div>
              </div>
              <div class="w-full px-1 bg-white py-5 lg:px-2 lg:py-2 tracking-wide">
                  <div class="flex flex-row lg:justify-start justify-center">
                      <div class="text-gray-700 font-light text-sm text-center lg:text-left px-2">
                          1:30 PM
                      </div>
                      <div class="text-gray-700 font-light text-sm text-center lg:text-left px-2">
                          Organiser : IHC
                      </div>
                  </div>
                  <div class="text-gray-700 text-xl pb-1 text-center lg:text-left px-2">
                      International Conference Dubai
                  </div>

                  <div class="text-gray-800 font-light text-sm pt-1 text-center lg:text-left px-2">
                      A-142/1, A-142, Ganesh Nagar, Tilak Nagar, New Delhi, 110018
                  </div>
              </div>
              <div class="flex flex-row items-center w-full lg:w-1/3 bg-white lg:justify-end justify-center px-2 py-4 lg:px-0 rounded-lg">
                  <span class="tracking-wider text-gray-600 cursor-pointer bg-gray-100 hover:bg-gray-200 px-4 py-2 text-sm rounded-lg leading-loose mx-2">
                      Going
                  </span>
              </div>
          </div>
      </div>
  </body>
  </html>`;
  const inputPath = path.join(folder, 'index.html');
  outputPath = path.join(folder, outputPath);
  fs.writeFileSync(inputPath, template);
  fs.writeFileSync(outputPath, '');
  return { html: inputPath, css: outputPath };
}
