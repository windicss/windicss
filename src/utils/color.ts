import colorString from 'color-string';
import type { Color } from 'color-string';

export function hsl2rgb(h: number, s: number, l: number): [number, number, number] {
  l /= 100;
  if (h >= 360) h %= 360;

  const c = (1 - Math.abs(2 * l - 1)) * (s / 100);
  const x = c * (1 - Math.abs((h / 60) % 2 - 1));
  const m = l - c/2;
  let r = 0;
  let g = 0;
  let b = 0;

  if (0 <= h && h < 60) {
    r = c + m; g = x + m; b = m;
  } else if (60 <= h && h < 120) {
    r = x + m; g = c + m; b = m;
  } else if (120 <= h && h < 180) {
    r = m; g = c + m; b = x + m;
  } else if (180 <= h && h < 240) {
    r = m; g = x + m; b = c + m;
  } else if (240 <= h && h < 300) {
    r = x + m; g = m; b = c + m;
  } else if (300 <= h && h < 360) {
    r = c + m; g = m; b = x + m;
  }

  return [
    Math.round(r * 255),
    Math.round(g * 255),
    Math.round(b * 255),
  ];
}

export function hwb2rgb(h: number, w: number, b: number): [number, number, number] {
  const rgb = hsl2rgb(h, 100, 50);

  for (let i = 0; i < 3; ++i) {
    let c = rgb[i] / 255;

    c *= 1 - w/100 - b/100;
    c += w/100;

    rgb[i] = Math.round(c * 255);
  }

  return rgb;
}

export function toRGBA(color: string): Color | undefined {
  if (/^hsla?/.test(color)) {
    const color_array = colorString.get.hsl(color);
    if (!color_array) return;
    return [...hsl2rgb(color_array[0], color_array[1], color_array[2]), color_array[3]];
  } else if (/^rgba?/.test(color)) {
    const color_array = colorString.get.rgb(color);
    if (!color_array) return;
    return color_array;
  } else if (color.startsWith('hwb')) {
    const color_array = colorString.get.hwb(color);
    if (!color_array) return;
    return [...hwb2rgb(color_array[0], color_array[1], color_array[2]), color_array[3]];
  }
  return colorString.get(color)?.value;
}

export function toRGB(color: string): number[] | undefined {
  const rgba = toRGBA(color);
  if (!rgba) return;
  rgba.pop();
  return rgba;
}

export function toColor(color_string: string) : { color: string, opacity: string } {
  const rgba = toRGBA(color_string);
  const color = rgba ? rgba.slice(0, 3).join(', ') : color_string;
  const opacity = rgba ? rgba[3].toString() : '1';

  return { color, opacity };
}
