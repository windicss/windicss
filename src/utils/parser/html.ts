export type Attr = { raw: string, key: string, value: string|string[], start: number, end: number };
export type ClassName = { result: string, start: number, end: number };

export default class HTMLParser {
  html?: string;
  constructor(html?: string) {
    this.html = html;
  }

  parseAttrs(): Attr[] {
    if (!this.html) return [];
    const output: Attr[] = [];
    const regex = /\S+\s*=\s*"[^"]+"|\S+\s*=\s*'[^']+'|\S+\s*=\s*[^>\s]+/igm;
    let match;
    while ((match = regex.exec(this.html as string))) {
      if (match) {
        const raw = match[0];
        const sep = raw.indexOf('=');
        const key = raw.slice(0, sep).trim();
        let value: string| string[] = raw.slice(sep + 1).trim();
        if (['"', '\''].includes(value.charAt(0))) value = value.slice(1, -1);
        value = value.split(/\s/).filter(i => i);
        value = value[0] === undefined ? '' : value[1] === undefined ? value[0] : value;
        output.push({
          raw,
          key,
          value,
          start: match.index,
          end: regex.lastIndex,
        });
      }
    }
    return output;
  }

  parseClasses(): ClassName[] {
    // Match all class properties
    if (!this.html) return [];
    const output: ClassName[] = [];
    const regex = /class(Name)?\s*=\s*{`[^`]+`}|class(Name)?\s*=\s*"[^"]+"|class(Name)?\s*=\s*'[^']+'|class(Name)?\s*=\s*[^>\s]+/igm;
    let match;
    while ((match = regex.exec(this.html as string))) {
      if (match) {
        const raw = match[0];
        const sep = raw.indexOf('=');
        let value: string| string[] = raw.slice(sep + 1).trim();
        let start = match.index + sep + 1 + (this.html.slice(sep + 1).match(/[^'"]/)?.index ?? 0);
        let end = regex.lastIndex;
        let first = value.charAt(0);
        while (['"', '\'', '`', '{'].includes(first)) {
          value = value.slice(1, -1);
          first = value.charAt(0);
          end--;
          start++;
        }
        output.push({
          result: value,
          start,
          end,
        });
      }
    }
    return output;
  }

  parseTags(): string[] {
    // Match all html tags
    if (!this.html) return [];
    return Array.from(new Set(this.html.match(/<\w+/g))).map((i) =>
      i.substring(1)
    );
  }
}
