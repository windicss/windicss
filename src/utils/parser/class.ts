import type { Element } from '../../interfaces';
import { isSpace } from '../../utils/tools';

export default class ClassParser {
  index: number;
  separator: string;
  variants: string[];
  classNames?: string;

  constructor(classNames?: string, separator = ':', variants?: string[]) {
    this.classNames = classNames;
    this.separator = separator;
    this.variants = variants || [];
    this.index = 0;
  }

  private _handle_group(removeDuplicated = true): Element[] {
    if (!this.classNames) return [];
    let preChar;
    let char;
    let group;
    let func;
    let variant;
    let variants = [];
    let variantStart = this.index + 1;
    let classStart = this.index + 1;
    let groupStart = this.index + 1;
    let important = false;
    let ignoreSpace = false;
    let ignoreBracket = false;
    let insideSquareBracket = false;
    let brackets = 0;
    const sepLength = this.separator.length;
    const parts: Element[] = [];
    const length = this.classNames.length;

    while (this.index < length) {
      this.index++;
      char = this.classNames.charAt(this.index);
      // ignore parsing and leave content inside square brackets as-is
      if (insideSquareBracket) {
        if(' \n\t\r'.includes(char)) {
          insideSquareBracket = false;
        } else {
          if (char === ']')
            insideSquareBracket = false;
          continue;
        }
      }
      // handle chars
      switch (char) {
      case '!':
        important = true;
        break;
      case this.separator[0]:
        if (this.classNames.slice(this.index, this.index + sepLength) === this.separator) {
          variant = this.classNames.slice(variantStart, this.index);
          if (variant.charAt(0) === '!') variant = variant.slice(1,);
          if (this.variants.includes(variant)) {
            variants.push(variant);
            this.index += sepLength - 1;
            variantStart = this.index + 1;
            ignoreSpace = true;
          }
        }
        break;
      case '[':
        insideSquareBracket = true;
        break;
      case '(':
        preChar = this.classNames.charAt(this.index - 1);
        if (preChar === '-' || (!ignoreSpace && preChar === ' ')) {
          ignoreBracket = true;
        } else if (ignoreSpace) {
          group = this._handle_group();
        } else {
          brackets += 1;
          func = this.classNames.slice(groupStart, this.index);
          while (!isSpace(this.classNames.charAt(this.index))) {
            this.index++;
          }
          this.index--;
        }
        ignoreSpace = false;
        break;
      case '"':
      case '`':
      case '\'':
      case ')':
      case ' ':
      case '\n':
      case '\t':
      case '\r':
        if (!ignoreSpace) {
          if (groupStart !== this.index) {
            const raw = this.classNames.slice(classStart, this.index);
            const start = classStart - 1;
            const end = this.index - 1;
            if (Array.isArray(group)) {
              parts.push({ raw, start, end, variants, content: group, type: 'group', important });
              group = undefined;
            } else if (func) {
              const utility = this.classNames.slice(variantStart, this.index);
              parts.push({ raw: raw, start, end, variants, content: utility, type: 'utility', important });
              func = undefined;
            } else if (ignoreBracket && char === ')') {
              // utility with bracket
              const utility = this.classNames.slice(variantStart, this.index + 1);
              parts.push({ raw: raw + ')', start, end: this.index, variants, content: important ? utility.slice(1,): utility, type: 'utility', important });
            } else {
              const utility = this.classNames.slice(variantStart, this.index);
              if (utility.charAt(0) === '*') {
                parts.push({ raw, start, end, variants, content: utility.slice(1,), type: 'alias', important });
              } else {
                parts.push({ raw, start, end, variants, content: utility.charAt(0) === '!' ? utility.slice(1,): utility, type: 'utility', important });
              }
            }
            variants = [];
            important = false;
          }
          groupStart = this.index + 1;
          classStart = this.index + 1;
        }
        variantStart = this.index + 1;
        break;
      default:
        ignoreSpace = false;
      }
      if (char === ')') {
        brackets -= 1;
        if (!ignoreBracket && brackets < 0) break; // end group
        ignoreBracket = false;
      }
    }

    if (removeDuplicated) {
      const newParts: Element[] = [];
      const cache: string[] = [];
      parts.forEach((item) => {
        if (!cache.includes(item.raw)) {
          cache.push(item.raw);
          newParts.push(item);
        }
      });
      return newParts;
    }
    return parts;
  }

  parse(removeDuplicated = true): Element[] {
    if (!this.classNames) return [];
    // Turn classes into group;
    this.classNames = '(' + this.classNames + ')';
    const elements = this._handle_group(removeDuplicated);
    // Initialization, convenient for next call
    this.index = 0;
    this.classNames = this.classNames.slice(1, -1);
    return elements;
  }
}
