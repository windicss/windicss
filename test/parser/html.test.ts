import { readFileSync } from 'fs';
import { HTMLParser } from '../../src/utils/parser';

const EXAMPLE_1 = readFileSync('./test/assets/example.html').toString();
const EXAMPLE_2 = readFileSync('./test/assets/example2.html').toString();
const EXAMPLE_3 = readFileSync('./test/assets/example3.razor').toString();

describe('HTMLParser', () => {
  it('parse tags', () => {
    const parser = new HTMLParser();
    expect(parser.parseTags().length).toBe(0);
    parser.html = EXAMPLE_1;
    expect(parser.parseTags()).toMatchSnapshot('tag');
  });

  it('parse classes', () => {
    const parser = new HTMLParser();
    expect(parser.parseClasses().length).toBe(0);
    parser.html = EXAMPLE_1;
    expect(parser.parseClasses()).toMatchSnapshot('class');
  });

  it('parse razor classes', () => {
    const parser = new HTMLParser();
    expect(parser.parseClasses().length).toBe(0);
    parser.html = EXAMPLE_3;
    expect(parser.parseClasses()).toMatchSnapshot('parse razor classes');
  });

  it('parse attributes', () => {
    const parser = new HTMLParser(EXAMPLE_2);
    expect(parser.parseAttrs()).toMatchSnapshot('attr');
  });

  it('parse jsx syntax', () => {
    const parser = new HTMLParser('class={`z-20 bg-green-600 ${variable}`}');
    expect(parser.parseClasses()).toMatchSnapshot('class');
    parser.html = '<div className={css`height:calc(100% - 56px)`}><div className="float-left w-250px h-full bg-hex-f7f7fa overflow-y-auto"></div></div>';
    expect(parser.parseClasses()).toMatchSnapshot('className');
  });

  it('parse razor syntax', () => {
    const parser = new HTMLParser('class=@(active ? "bg-green-600" : "bg-red-500")');
    expect(parser.parseClasses()).toMatchSnapshot('class');
  });
});
