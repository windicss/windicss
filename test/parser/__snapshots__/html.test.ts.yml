Tools / parse attributes / attr / 0: |-
  [
    {
      "raw": "container=\"~\"",
      "key": "container",
      "value": "~",
      "start": 5,
      "end": 18
    },
    {
      "raw": "min-h=\"screen\"",
      "key": "min-h",
      "value": "screen",
      "start": 19,
      "end": 33
    },
    {
      "raw": "bg=\"gray-100\"",
      "key": "bg",
      "value": "gray-100",
      "start": 34,
      "end": 47
    },
    {
      "raw": "flex=\"~ col\"",
      "key": "flex",
      "value": [
        "~",
        "col"
      ],
      "start": 48,
      "end": 60
    },
    {
      "raw": "justify=\"center\"",
      "key": "justify",
      "value": "center",
      "start": 61,
      "end": 77
    },
    {
      "raw": "p=\"y-6 sm:y-12 md:y-0 lg:y-2 xl:y-4\"",
      "key": "p",
      "value": [
        "y-6",
        "sm:y-12",
        "md:y-0",
        "lg:y-2",
        "xl:y-4"
      ],
      "start": 78,
      "end": 114
    },
    {
      "raw": "pos=\"relative\"",
      "key": "pos",
      "value": "relative",
      "start": 123,
      "end": 137
    },
    {
      "raw": "w=\"sm:max-xl\"",
      "key": "w",
      "value": "sm:max-xl",
      "start": 138,
      "end": 151
    },
    {
      "raw": "p=\"y-3\"",
      "key": "p",
      "value": "y-3",
      "start": 152,
      "end": 159
    },
    {
      "raw": "m=\"sm:x-auto\"",
      "key": "m",
      "value": "sm:x-auto",
      "start": 160,
      "end": 173
    },
    {
      "raw": "pos=\"absolute inset-0\"",
      "key": "pos",
      "value": [
        "absolute",
        "inset-0"
      ],
      "start": 184,
      "end": 206
    },
    {
      "raw": "gradient=\"to-r from-cyan-400 to-light-blue-500\"",
      "key": "gradient",
      "value": [
        "to-r",
        "from-cyan-400",
        "to-light-blue-500"
      ],
      "start": 207,
      "end": 254
    },
    {
      "raw": "shadow=\"lg\"",
      "key": "shadow",
      "value": "lg",
      "start": 255,
      "end": 266
    },
    {
      "raw": "transform=\"~ -skew-y-6 sm:skew-y-0 sm:-rotate-6\"",
      "key": "transform",
      "value": [
        "~",
        "-skew-y-6",
        "sm:skew-y-0",
        "sm:-rotate-6"
      ],
      "start": 267,
      "end": 315
    },
    {
      "raw": "border=\"sm:rounded-3xl\"",
      "key": "border",
      "value": "sm:rounded-3xl",
      "start": 316,
      "end": 339
    },
    {
      "raw": "pos=\"relative\"",
      "key": "pos",
      "value": "relative",
      "start": 361,
      "end": 375
    },
    {
      "raw": "p=\"x-4 y-10 sm:20\"",
      "key": "p",
      "value": [
        "x-4",
        "y-10",
        "sm:20"
      ],
      "start": 376,
      "end": 394
    },
    {
      "raw": "bg=\"white\"",
      "key": "bg",
      "value": "white",
      "start": 395,
      "end": 405
    },
    {
      "raw": "shadow=\"lg\"",
      "key": "shadow",
      "value": "lg",
      "start": 406,
      "end": 417
    },
    {
      "raw": "border=\"sm:rounded-3xl\"",
      "key": "border",
      "value": "sm:rounded-3xl",
      "start": 418,
      "end": 441
    },
    {
      "raw": "m=\"x-auto\"",
      "key": "m",
      "value": "x-auto",
      "start": 454,
      "end": 464
    },
    {
      "raw": "w=\"max-md\"",
      "key": "w",
      "value": "max-md",
      "start": 465,
      "end": 475
    },
    {
      "raw": "src=\"/img/logo.svg\"",
      "key": "src",
      "value": "/img/logo.svg",
      "start": 506,
      "end": 525
    },
    {
      "raw": "h=\"7 sm:8\"",
      "key": "h",
      "value": [
        "7",
        "sm:8"
      ],
      "start": 526,
      "end": 536
    },
    {
      "raw": "divide=\"y gray-200\"",
      "key": "divide",
      "value": [
        "y",
        "gray-200"
      ],
      "start": 567,
      "end": 586
    },
    {
      "raw": "text=\"base sm:lg gray-700\"",
      "key": "text",
      "value": [
        "base",
        "sm:lg",
        "gray-700"
      ],
      "start": 603,
      "end": 629
    },
    {
      "raw": "font=\"leading-6 sm:leading-7\"",
      "key": "font",
      "value": [
        "leading-6",
        "sm:leading-7"
      ],
      "start": 630,
      "end": 659
    },
    {
      "raw": "p=\"y-8\"",
      "key": "p",
      "value": "y-8",
      "start": 660,
      "end": 667
    },
    {
      "raw": "space=\"y-4\"",
      "key": "space",
      "value": "y-4",
      "start": 668,
      "end": 679
    },
    {
      "raw": "list=\"disc\"",
      "key": "list",
      "value": "disc",
      "start": 796,
      "end": 807
    },
    {
      "raw": "space=\"y-2\"",
      "key": "space",
      "value": "y-2",
      "start": 808,
      "end": 819
    },
    {
      "raw": "flex=~",
      "key": "flex",
      "value": "~",
      "start": 839,
      "end": 845
    },
    {
      "raw": "items=\"start\"",
      "key": "items",
      "value": "start",
      "start": 846,
      "end": 859
    },
    {
      "raw": "h=\"6 sm:7\"",
      "key": "h",
      "value": [
        "6",
        "sm:7"
      ],
      "start": 883,
      "end": 893
    },
    {
      "raw": "flex=\"~\"",
      "key": "flex",
      "value": "~",
      "start": 894,
      "end": 902
    },
    {
      "raw": "items=\"center\"",
      "key": "items",
      "value": "center",
      "start": 903,
      "end": 917
    },
    {
      "raw": "flex=\"shrink-0\"",
      "key": "flex",
      "value": "shrink-0",
      "start": 942,
      "end": 957
    },
    {
      "raw": "h=5",
      "key": "h",
      "value": "5",
      "start": 958,
      "end": 961
    },
    {
      "raw": "w=5",
      "key": "w",
      "value": "5",
      "start": 962,
      "end": 965
    },
    {
      "raw": "text=\"cyan-500\"",
      "key": "text",
      "value": "cyan-500",
      "start": 966,
      "end": 981
    },
    {
      "raw": "viewBox=\"0 0 20 20\"",
      "key": "viewBox",
      "value": [
        "0",
        "0",
        "20",
        "20"
      ],
      "start": 982,
      "end": 1001
    },
    {
      "raw": "fill=\"currentColor\"",
      "key": "fill",
      "value": "currentColor",
      "start": 1002,
      "end": 1021
    },
    {
      "raw": "fill-rule=\"evenodd\"",
      "key": "fill-rule",
      "value": "evenodd",
      "start": 1049,
      "end": 1068
    },
    {
      "raw": "d=\"M10 18a8 8 0 100-16 8 8 0 000 16zm3.707-9.293a1 1 0 00-1.414-1.414L9 10.586 7.707 9.293a1 1 0 00-1.414 1.414l2 2a1 1 0 001.414 0l4-4z\"",
      "key": "d",
      "value": [
        "M10",
        "18a8",
        "8",
        "0",
        "100-16",
        "8",
        "8",
        "0",
        "000",
        "16zm3.707-9.293a1",
        "1",
        "0",
        "00-1.414-1.414L9",
        "10.586",
        "7.707",
        "9.293a1",
        "1",
        "0",
        "00-1.414",
        "1.414l2",
        "2a1",
        "1",
        "0",
        "001.414",
        "0l4-4z"
      ],
      "start": 1091,
      "end": 1228
    },
    {
      "raw": "clip-rule=\"evenodd\"",
      "key": "clip-rule",
      "value": "evenodd",
      "start": 1251,
      "end": 1270
    },
    {
      "raw": "class=\"ml-2\"",
      "key": "class",
      "value": "ml-2",
      "start": 1342,
      "end": 1354
    },
    {
      "raw": "class=\"text-sm font-bold text-gray-900\"",
      "key": "class",
      "value": [
        "text-sm",
        "font-bold",
        "text-gray-900"
      ],
      "start": 1415,
      "end": 1454
    },
    {
      "raw": "flex=\"~\"",
      "key": "flex",
      "value": "~",
      "start": 1542,
      "end": 1550
    },
    {
      "raw": "items=\"start\"",
      "key": "items",
      "value": "start",
      "start": 1551,
      "end": 1564
    },
    {
      "raw": "flex=\"~\"",
      "key": "flex",
      "value": "~",
      "start": 1588,
      "end": 1596
    },
    {
      "raw": "items=\"center\"",
      "key": "items",
      "value": "center",
      "start": 1597,
      "end": 1611
    },
    {
      "raw": "h=\"6 sm:7\"",
      "key": "h",
      "value": [
        "6",
        "sm:7"
      ],
      "start": 1612,
      "end": 1622
    },
    {
      "raw": "flex=\"shrink-0\"",
      "key": "flex",
      "value": "shrink-0",
      "start": 1647,
      "end": 1662
    },
    {
      "raw": "h=5",
      "key": "h",
      "value": "5",
      "start": 1663,
      "end": 1666
    },
    {
      "raw": "w=5",
      "key": "w",
      "value": "5",
      "start": 1667,
      "end": 1670
    },
    {
      "raw": "text=cyan-500",
      "key": "text",
      "value": "cyan-500",
      "start": 1671,
      "end": 1684
    },
    {
      "raw": "viewBox=\"0 0 20 20\"",
      "key": "viewBox",
      "value": [
        "0",
        "0",
        "20",
        "20"
      ],
      "start": 1685,
      "end": 1704
    },
    {
      "raw": "fill=\"currentColor\"",
      "key": "fill",
      "value": "currentColor",
      "start": 1705,
      "end": 1724
    },
    {
      "raw": "fill-rule=\"evenodd\"",
      "key": "fill-rule",
      "value": "evenodd",
      "start": 1752,
      "end": 1771
    },
    {
      "raw": "d=\"M10 18a8 8 0 100-16 8 8 0 000 16zm3.707-9.293a1 1 0 00-1.414-1.414L9 10.586 7.707 9.293a1 1 0 00-1.414 1.414l2 2a1 1 0 001.414 0l4-4z\"",
      "key": "d",
      "value": [
        "M10",
        "18a8",
        "8",
        "0",
        "100-16",
        "8",
        "8",
        "0",
        "000",
        "16zm3.707-9.293a1",
        "1",
        "0",
        "00-1.414-1.414L9",
        "10.586",
        "7.707",
        "9.293a1",
        "1",
        "0",
        "00-1.414",
        "1.414l2",
        "2a1",
        "1",
        "0",
        "001.414",
        "0l4-4z"
      ],
      "start": 1794,
      "end": 1931
    },
    {
      "raw": "clip-rule=\"evenodd\"",
      "key": "clip-rule",
      "value": "evenodd",
      "start": 1954,
      "end": 1973
    },
    {
      "raw": "m=l-2",
      "key": "m",
      "value": "l-2",
      "start": 2045,
      "end": 2050
    },
    {
      "raw": "font=\"bold\"",
      "key": "font",
      "value": "bold",
      "start": 2118,
      "end": 2129
    },
    {
      "raw": "text=\"sm gray-900\"",
      "key": "text",
      "value": [
        "sm",
        "gray-900"
      ],
      "start": 2130,
      "end": 2148
    },
    {
      "raw": "flex=\"~\"",
      "key": "flex",
      "value": "~",
      "start": 2222,
      "end": 2230
    },
    {
      "raw": "items=\"start\"",
      "key": "items",
      "value": "start",
      "start": 2231,
      "end": 2244
    },
    {
      "raw": "flex=\"~\"",
      "key": "flex",
      "value": "~",
      "start": 2268,
      "end": 2276
    },
    {
      "raw": "items=\"center\"",
      "key": "items",
      "value": "center",
      "start": 2277,
      "end": 2291
    },
    {
      "raw": "h=\"6 sm:7\"",
      "key": "h",
      "value": [
        "6",
        "sm:7"
      ],
      "start": 2292,
      "end": 2302
    },
    {
      "raw": "class=\"flex-shrink-0 h-5 w-5 text-cyan-500\"",
      "key": "class",
      "value": [
        "flex-shrink-0",
        "h-5",
        "w-5",
        "text-cyan-500"
      ],
      "start": 2327,
      "end": 2370
    },
    {
      "raw": "viewBox=\"0 0 20 20\"",
      "key": "viewBox",
      "value": [
        "0",
        "0",
        "20",
        "20"
      ],
      "start": 2371,
      "end": 2390
    },
    {
      "raw": "fill=\"currentColor\"",
      "key": "fill",
      "value": "currentColor",
      "start": 2391,
      "end": 2410
    },
    {
      "raw": "fill-rule=\"evenodd\"",
      "key": "fill-rule",
      "value": "evenodd",
      "start": 2438,
      "end": 2457
    },
    {
      "raw": "d=\"M10 18a8 8 0 100-16 8 8 0 000 16zm3.707-9.293a1 1 0 00-1.414-1.414L9 10.586 7.707 9.293a1 1 0 00-1.414 1.414l2 2a1 1 0 001.414 0l4-4z\"",
      "key": "d",
      "value": [
        "M10",
        "18a8",
        "8",
        "0",
        "100-16",
        "8",
        "8",
        "0",
        "000",
        "16zm3.707-9.293a1",
        "1",
        "0",
        "00-1.414-1.414L9",
        "10.586",
        "7.707",
        "9.293a1",
        "1",
        "0",
        "00-1.414",
        "1.414l2",
        "2a1",
        "1",
        "0",
        "001.414",
        "0l4-4z"
      ],
      "start": 2480,
      "end": 2617
    },
    {
      "raw": "clip-rule=\"evenodd\"",
      "key": "clip-rule",
      "value": "evenodd",
      "start": 2640,
      "end": 2659
    },
    {
      "raw": "m=\"l-2\"",
      "key": "m",
      "value": "l-2",
      "start": 2731,
      "end": 2738
    },
    {
      "raw": "text=\"base sm:lg\"",
      "key": "text",
      "value": [
        "base",
        "sm:lg"
      ],
      "start": 2988,
      "end": 3005
    },
    {
      "raw": "p=\"t-6\"",
      "key": "p",
      "value": "t-6",
      "start": 3006,
      "end": 3013
    },
    {
      "raw": "font=\"bold leading-6 sm:leading-7\"",
      "key": "font",
      "value": [
        "bold",
        "leading-6",
        "sm:leading-7"
      ],
      "start": 3014,
      "end": 3048
    },
    {
      "raw": "href=\"https://windicss.org/utilities\"",
      "key": "href",
      "value": "https://windicss.org/utilities",
      "start": 3133,
      "end": 3170
    },
    {
      "raw": "text=\"cyan-600 hover:cyan-700\"",
      "key": "text",
      "value": [
        "cyan-600",
        "hover:cyan-700"
      ],
      "start": 3171,
      "end": 3201
    }
  ]
Tools / parse classes / class / 0: |-
  [
    {
      "result": "container min-h-screen bg-gray-100 py-6 flex flex-col justify-center sm:py-12 md:py-0 lg:py-2 xl:py-4",
      "start": 12,
      "end": 113
    },
    {
      "result": "relative py-3 sm:max-w-xl sm:mx-auto",
      "start": 130,
      "end": 166
    },
    {
      "result": "absolute inset-0 bg-gradient-to-r from-cyan-400 to-light-blue-500 shadow-lg transform -skew-y-6 sm:skew-y-0 sm:-rotate-6 sm:rounded-3xl",
      "start": 185,
      "end": 320
    },
    {
      "result": "relative px-4 py-10 bg-white shadow-lg sm:rounded-3xl sm:p-20",
      "start": 350,
      "end": 411
    },
    {
      "result": "max-w-md mx-auto",
      "start": 432,
      "end": 448
    },
    {
      "result": "h-7 sm:h-8",
      "start": 507,
      "end": 517
    },
    {
      "result": "divide-y divide-gray-200",
      "start": 557,
      "end": 581
    },
    {
      "result": "py-8 text-base leading-6 space-y-4 text-gray-700 sm:text-lg sm:leading-7",
      "start": 606,
      "end": 678
    },
    {
      "result": "list-disc space-y-2",
      "start": 803,
      "end": 822
    },
    {
      "result": "flex items-start",
      "start": 850,
      "end": 866
    },
    {
      "result": "h-6 flex items-center sm:h-7",
      "start": 898,
      "end": 926
    },
    {
      "result": "flex-shrink-0 h-5 w-5 text-cyan-500",
      "start": 959,
      "end": 994
    },
    {
      "result": "ml-2",
      "start": 1363,
      "end": 1367
    },
    {
      "result": "text-sm font-bold text-gray-900",
      "start": 1436,
      "end": 1467
    },
    {
      "result": "flex items-start",
      "start": 1563,
      "end": 1579
    },
    {
      "result": "h-6 flex items-center sm:h-7",
      "start": 1611,
      "end": 1639
    },
    {
      "result": "flex-shrink-0 h-5 w-5 text-cyan-500",
      "start": 1672,
      "end": 1707
    },
    {
      "result": "ml-2",
      "start": 2076,
      "end": 2080
    },
    {
      "result": "text-sm font-bold text-gray-900",
      "start": 2156,
      "end": 2187
    },
    {
      "result": "flex items-start",
      "start": 2269,
      "end": 2285
    },
    {
      "result": "h-6 flex items-center sm:h-7",
      "start": 2317,
      "end": 2345
    },
    {
      "result": "flex-shrink-0 h-5 w-5 text-cyan-500",
      "start": 2378,
      "end": 2413
    },
    {
      "result": "ml-2",
      "start": 2781,
      "end": 2785
    },
    {
      "result": "pt-6 text-base leading-6 font-bold sm:text-lg sm:leading-7",
      "start": 3042,
      "end": 3100
    },
    {
      "result": "text-cyan-600 hover:text-cyan-700",
      "start": 3231,
      "end": 3264
    }
  ]
Tools / parse jsx syntax / class / 0: |-
  [
    {
      "result": "z-20 bg-green-600 ${variable}",
      "start": 8,
      "end": 37
    }
  ]
Tools / parse jsx syntax / className / 1: |-
  [
    {
      "result": "css`height:calc(100",
      "start": 16,
      "end": 35
    },
    {
      "result": "float-left w-250px h-full bg-hex-f7f7fa overflow-y-auto",
      "start": 63,
      "end": 118
    }
  ]
Tools / parse tags / tag / 0: |-
  [
    "div",
    "img",
    "p",
    "ul",
    "li",
    "span",
    "svg",
    "path",
    "code",
    "a"
  ]
